### PR TITLE
Test case for scalling down before scale up finished

### PR DESCRIPTION
**Pull request by dshulyak**
Verifies that current pet (one which was created last by scale up action)
will enter running before scale down will take place

related: #30082

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34454)

### CODE REVIEW

**Comment by dshulyak** [comment]
Includes https://github.com/kubernetes/kubernetes/pull/32868

@bprashanth PTAL

**Comment by erictune** [comment]
@janetkuo PTAL

File: test/e2e/petset.go [hunk 1]
```diff
@@ -177,6 +209,175 @@ var _ = framework.KubeDescribe("PetSet [Slow] [Feature:PetSet]", func() {
...
+			pst.update(ps.Namespace, ps.Name, func(ps *apps.PetSet) { ps.Spec.Replicas = 1 })
+			pst.restoreProbe(ps, testProbe)
+			for i := len(podTracker) - 1; i >= 0; i-- {
+				event := podTracker[i]
+				if event.pod.Name != lastPet {
+					continue
+				}
+				isReady := api.IsPodReady(event.pod)
+				if event.pod.Status.Phase != api.PodRunning || !isReady {
+					framework.Failf("%v expected to be running != %v and ready != %v", event.pod.Name, event.pod.Status.Phase, isReady)
```
Comment (janetkuo) [inline test/e2e/petset.go]: since it's OR, it not necessary be "the running != %v and ready != %v"

File: test/e2e/petset.go [hunk 2]
```diff
@@ -177,6 +209,175 @@ var _ = framework.KubeDescribe("PetSet [Slow] [Feature:PetSet]", func() {
...
+			pst.setHealthy(ps)
+			pst.update(ps.Namespace, ps.Name, func(ps *apps.PetSet) { ps.Spec.Replicas = 3 })
+			pst.waitForRunningAndReady(2, ps)
+			pst.breakProbe(ps, testProbe)
+			pst.waitForRunningAndNotReady(2, ps)
+			pst.setHealthy(ps)
+			pst.update(ps.Namespace, ps.Name, func(ps *apps.PetSet) { ps.Spec.Replicas = 1 })
+			pst.restoreProbe(ps, testProbe)
+			for i := len(podTracker) - 1; i >= 0; i-- {
+				event := podTracker[i]
```
Comment (janetkuo) [inline test/e2e/petset.go]: Could you explain more on what this loop is doing? Is it looking at the last (and only the last) event on `pet-2` to see if the pod is ready? Do we need to wait before the check?

**Comment by bprashanth** [comment]
@janetkuo thanks! let me know if you need help, otherwise I'll assume you got this

**Comment by janetkuo** [comment]
Ping @dshulyak for addressing comments

File: test/e2e/petset.go (hunk 2 above)
Comment (dshulyak) [inline test/e2e/petset.go]: i wanted to verify that lastPet entered running before scale down started. Will rewrite with watcher

Comment (dshulyak) [inline test/e2e/petset.go]: and it is in opposite direction because i wanted to find last event for this pod

**Comment by dshulyak** [comment]
@janetkuo fixed comments and rebased, please take a look when you will have time

File: test/e2e/petset.go [hunk 3]
```diff
@@ -212,6 +212,64 @@ var _ = framework.KubeDescribe("StatefulSet [Slow] [Feature:PetSet]", func() {
...
+			By("scaling up stateful set " + psName + " to 3 replicas")
+			pst.setHealthy(ps)
+			pst.updateReplicas(ps, 3)
+			pst.waitForRunningAndReady(2, ps)
+
+			By("before scale up finished setting 2nd pod to be not ready by breaking readiness probe")
+			pst.breakProbe(ps, testProbe)
+			pst.waitForRunningAndNotReady(2, ps)
+
+			By("scaling down to 1 replica and restoring health of 2nd pod")
```
Comment (janetkuo) [inline test/e2e/petset.go]: "Continue scale operations after the 2nd pod, and scaling down to 1 replica"

File: test/e2e/petset.go [hunk 4]
```diff
@@ -212,6 +212,64 @@ var _ = framework.KubeDescribe("StatefulSet [Slow] [Feature:PetSet]", func() {
 			pst.verifyPodAtIndex(updateIndex, ps, verify)
 		})
 
+		It("scaling down before scale up is finished should wait until current pod will be running and ready before it will be removed", func() {
+			By("creating petset " + psName + " in namespace " + ns)
```
Comment (janetkuo) [inline test/e2e/petset.go]: `"Creating stateful set " + psName + " in namespace " + ns + ", and pausing scale operations after each pod"`

File: test/e2e/petset.go [hunk 5]
```diff
@@ -212,6 +212,64 @@ var _ = framework.KubeDescribe("StatefulSet [Slow] [Feature:PetSet]", func() {
...
+				Path: "/index.html",
+				Port: intstr.IntOrString{IntVal: 80}}}}
+			ps := newStatefulSet(psName, ns, headlessSvcName, 1, nil, nil, labels, testProbe)
+			setInitializedAnnotation(ps, "false")
+			ps, err := c.Apps().StatefulSets(ns).Create(ps)
+			Expect(err).NotTo(HaveOccurred())
+			pst := &statefulSetTester{c: c}
+			pst.waitForRunningAndReady(1, ps)
+
+			By("scaling up stateful set " + psName + " to 3 replicas")
```
Comment (janetkuo) [inline test/e2e/petset.go]: ... and pausing after the 2nd pod

File: test/e2e/petset.go [hunk 6]
```diff
@@ -212,6 +212,64 @@ var _ = framework.KubeDescribe("StatefulSet [Slow] [Feature:PetSet]", func() {
...
+
+			By("before scale up finished setting 2nd pod to be not ready by breaking readiness probe")
+			pst.breakProbe(ps, testProbe)
+			pst.waitForRunningAndNotReady(2, ps)
+
+			By("scaling down to 1 replica and restoring health of 2nd pod")
+			pst.setHealthy(ps)
+			pst.updateReplicas(ps, 1)
+
+			By("verifying that 2nd pod will not be removed if it is not healthy")
```
Comment (janetkuo) [inline test/e2e/petset.go]: "Verifying that the 2nd pod will not be removed if it's not running and ready"

File: test/e2e/petset.go [hunk 7]
```diff
@@ -212,6 +212,64 @@ var _ = framework.KubeDescribe("StatefulSet [Slow] [Feature:PetSet]", func() {
...
+			watcher, err := f.ClientSet.Core().Pods(ns).Watch(api.SingleObject(
+				api.ObjectMeta{
+					Name:            expectedPod.Name,
+					ResourceVersion: expectedPod.ResourceVersion,
+				},
+			))
+			Expect(err).NotTo(HaveOccurred())
+			pst.restoreProbe(ps, testProbe)
+
+			By("verifying that 2nd pod will enter running and ready before it will be removed by scale down")
```
Comment (janetkuo) [inline test/e2e/petset.go]: "Verifying the 2nd pod is not removed by scaling down until it becomes running and ready"

File: test/e2e/petset.go [hunk 8]
```diff
@@ -212,6 +212,64 @@ var _ = framework.KubeDescribe("StatefulSet [Slow] [Feature:PetSet]", func() {
...
+			expectedPod, err := f.ClientSet.Core().Pods(ns).Get(expectedPodName)
+			Expect(err).NotTo(HaveOccurred())
+			watcher, err := f.ClientSet.Core().Pods(ns).Watch(api.SingleObject(
+				api.ObjectMeta{
+					Name:            expectedPod.Name,
+					ResourceVersion: expectedPod.ResourceVersion,
+				},
+			))
+			Expect(err).NotTo(HaveOccurred())
+			pst.restoreProbe(ps, testProbe)
```
Comment (janetkuo) [inline test/e2e/petset.go]: Move this down to the below `By`

**Comment by janetkuo** [comment]
Only nits, LGTM otherwise. This needs to wait for #32868

**Comment by dshulyak** [comment]
@k8s-bot kops aws e2e test this

**Comment by dshulyak** [comment]
@k8s-bot gci gke e2e test this

**Comment by dshulyak** [comment]
@k8s-bot cvm gke e2e test this

**Comment by dshulyak** [comment]
@k8s-bot kops aws e2e test this

**Review by janetkuo — APPROVED** [review]
/lgtm

**Comment by janetkuo** [comment]
Adding do-not-merge since this needs to wait for #32868

**Comment by janetkuo** [comment]
#32868 is merged

**Pull request merged**